src/python/grpcio_tests/tests_aio/unit/server_interceptor_test.py

```diff
@@ -162,6 +180,138 @@ async def test_apply_different_interceptors_by_metadata(self):
+                messages_pb2.SimpleResponse(payload=messages_pb2.Payload(
+                    body=b'\x42'))
+        }
+
+        async def intercept_and_cache(
+                continuation: Callable[[grpc.HandlerCallDetails], Awaitable[
+                    grpc.RpcMethodHandler]],
+                handler_call_details: grpc.HandlerCallDetails
+        ) -> grpc.RpcMethodHandler:
+            # Get the actual handler
+            handler = await continuation(handler_call_details)
+
+            def wrapper(behavior: Callable[
+                [messages_pb2.SimpleRequest, aio.
+                 ServerInterceptor], messages_pb2.SimpleResponse]):
+
+                @functools.wraps(behavior)
+                async def wrapper(request: messages_pb2.SimpleRequest,
+                                  context: aio.ServicerContext
+                                 ) -> messages_pb2.SimpleResponse:
+                    if request.response_size not in cache_store:
+                        cache_store[request.response_size] = await behavior(
+                            request, context)
+                    return cache_store[request.response_size]
+
+                return wrapper
+
+            return wrap_server_method_handler(wrapper, handler)
+
+        # Constructs a server with the cache interceptor
+        server, stub = await _create_server_stub_pair(
+            _GenericInterceptor(intercept_and_cache))
+
+        # Tests if the cache store is used
+        response = await stub.UnaryCall(
+            messages_pb2.SimpleRequest(response_size=42))
+        self.assertEqual(1, len(cache_store[42].payload.body))
+        self.assertEqual(cache_store[42], response)
+
+        # Tests response can be cached
+        response = await stub.UnaryCall(
+            messages_pb2.SimpleRequest(response_size=1337))
+        self.assertEqual(1337, len(cache_store[1337].payload.body))
+        self.assertEqual(cache_store[1337], response)
+        response = await stub.UnaryCall(
+            messages_pb2.SimpleRequest(response_size=1337))
+        self.assertEqual(cache_store[1337], response)
+
+    async def test_interceptor_unary_stream(self):
+        record = []
+        server, stub = await _create_server_stub_pair(
+            _LoggingInterceptor('log_unary_stream', record))
+
+        # Prepares the request
+        request = messages_pb2.StreamingOutputCallRequest()
+        for _ in range(_NUM_STREAM_RESPONSES):
+            request.response_parameters.append(
+                messages_pb2.ResponseParameters(size=_RESPONSE_PAYLOAD_SIZE,))
+
+        # Tests if the cache store is used
+        call = stub.StreamingOutputCall(request)
+
+        # Ensures the RPC goes fine
+        async for response in call:
+            self.assertEqual(_RESPONSE_PAYLOAD_SIZE, len(response.payload.body))
+        self.assertEqual(await call.code(), grpc.StatusCode.OK)
+
+        self.assertSequenceEqual([
+            'log_unary_stream:intercept_service',
```

Comment:
Should we provide a unit test that tests effectively the streaming arities, so being able to intercept the request or the response stream? For that, we would need to do the same that you already did for the unary call.

Reply:
Intercepting streaming request or response is quite difficult with this design. The interceptor author would need to duplicate the interface of ServicerContext, and synchronize it with a queue. Link to a TODO issue https://github.com/grpc/grpc/issues/18191 requesting for a complete server interceptor.

This PR proves server interceptors can intercept RPCs and act upon metadata, or unary requests. This is as-good-as existing stack, we can open another PR for possible ways to achieve functionalities better than existing stack.